aws/rust-runtime/aws-http/src/content_encoding.rs

```diff
@@ -0,0 +1,414 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use aws_smithy_checksums::body::ChecksumBody;
+use aws_smithy_http::body::SdkBody;
+
+use bytes::{Buf, Bytes, BytesMut};
+use http::{HeaderMap, HeaderValue};
+use http_body::{Body, SizeHint};
+use pin_project::pin_project;
+
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+const CRLF: &str = "\r\n";
+const CHUNK_TERMINATOR: &str = "0\r\n";
+
+/// Content encoding header value constants
+pub mod header_value {
+    /// Header value denoting "aws-chunked" encoding
+    pub const AWS_CHUNKED: &str = "aws-chunked";
+}
+
+/// Options used when constructing an [`AwsChunkedBody`][AwsChunkedBody].
+#[derive(Debug, Default)]
+#[non_exhaustive]
+pub struct AwsChunkedBodyOptions {
+    /// The total size of the stream. For unsigned encoding this implies that
+    /// there will only be a single chunk containing the underlying payload,
+    /// unless ChunkLength is also specified.
+    pub stream_length: Option<u64>,
+    /// The maximum size of each chunk to be sent.
+    ///
+    /// If ChunkLength and stream_length are both specified, the stream will be
+    /// broken up into chunk_length chunks. The encoded length of the aws-chunked
+    /// encoding can still be determined as long as all trailers, if any, have a
+    /// fixed length.
+    pub chunk_length: Option<u64>,
+    /// The length of each trailer sent within an `AwsChunkedBody`. Necessary in
+    /// order to correctly calculate the total size of the body accurately.
+    pub trailer_lens: Vec<u64>,
+}
```

Comment:
seems like this is tightly coupled with ChecksumBody—should it be?

Reply:
fixed